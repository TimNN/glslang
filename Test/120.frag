#version 120

float lowp;
float mediump;
float highp;

float precision;

in vec4 i;
out vec4 o;

uniform sampler2D s2D;
centroid varying vec2 centTexCoord;

uniform mat4x2 m;

struct s {
    float f;
};

void main()
{
    mat2x3 m23 = mat2x3(m);

    int a;
    bool b;
    s sv = s(a);
    float[2] ia = float[2](3, i.y);
    float f1 = 1;
    float f = a;
    f = a;
    ivec3 iv3;
    vec3 v3 = iv3;
    f = f + a;
    f = a - f;
    f += a;
    f = a - f;
    v3 *= iv3;
    v3 = iv3 / 2.0;
    v3 = 3.0 * iv3;
    v3 = 2 * v3;
    v3 = v3 - 2;
    if (f <  a ||
        a <= f ||
        f >  a ||
        f >= a ||
        a == f ||
        f != a);
    f = b ? a : f;
    f = b ? f : a;
    f = b ? a : a;
    s news = sv;
    
    i.xy + i.xyz;      // ERROR
    m * i.xyz;         // ERROR
    m + i;             // ERROR
    int aoeu = 1.0;    // ERROR
    f = b;             // ERROR
    f = a + b;         // ERROR
    f = b * a;         // ERROR
    b = a;             // ERROR
    b = b + f;         // ERROR
    f |= b;            // ERROR

    gl_FragColor = texture2D(s2D, centTexCoord);

    float flat;
    float smooth;
    float noperspective;
    float uvec2;
    float uvec3;
    float uvec4;
    //packed;     // ERROR, reserved word

    {
        mat4 m;
        vec4 v;
        bool b;
        gl_FragColor += b ? v : m;  // ERROR, types don't match around ":"
    }

    gl_FragColor.xr;    // ERROR, swizzlers not from same field space
    gl_FragColor.xyxyx.xy; // ERROR, cannot make a vec5, even temporarily
    centTexCoord.z;     // ERROR, swizzler out of range
    (a,b) = true;       // ERROR, not an l-value
}

float imageBuffer;
float uimage2DRect;

int main() {}           // ERROR
void main(int a) {}     // ERROR

const int a;            // ERROR

int foo(in float a);
int foo(out float a)    // ERROR
{
    return 3.2;         // ERROR
    foo(a);             // ERROR
}

bool gen(vec3 v)
{
    if (abs(v[0]) < 1e-4 && abs(v[1]) < 1e-4)
        return true;
}

void v1()
{
}

void v2()
{
    return v1();  // ERROR, no expression allowed, even though void
}
