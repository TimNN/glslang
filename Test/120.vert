#version 120

in vec4 i;                // ERROR
out vec4 o;               // ERROR

attribute vec2 attv2;
attribute vec4 attv4;
uniform sampler2D s2D;
invariant varying vec2 centTexCoord;
invariant gl_Position;
centroid gl_Position;     // ERROR
centroid centroid foo;    // ERROR
invariant gl_Position, gl_PointSize;

void main()
{
    centTexCoord = attv2; 
    gl_Position = attv4;

    vec3[12] a;
    vec4[a.length()] b;
    gl_Position = b[b.length()-1];

    float f[];
    int a1 = f.length();  // ERROR
    float f[7];
    int aa = f.length();
    int a2 = f.length;    // ERROR
    int a3 = f.length(a); // ERROR
    int a4 = f.flizbit;   // ERROR
    int a4 = f.flizbit(); // ERROR
    float md[2][4];       // ERROR
    float[2] md2[4];      // ERROR
    float[2][4] md3;      // ERROR
    float md5, md6[2][3]; // ERROR
    float[2] md4, md7[4]; // ERROR
    float md9[2][3] = float[2][3](1, 2, 3, 4, 5, 6);  // ERROR
    float md10, md11[2][3] = float[2][3](1, 2, 3, 4, 5, 6);  // ERROR

    gl_PointSize = 3.8;
}

uniform float initted = 3.4;   // okay

const float concall = sin(0.3);

int[2][3] foo(                 // ERROR
              float[2][3] a,   // ERROR
              float[2] b[3],   // ERROR
              float c[2][3]);  // ERROR
