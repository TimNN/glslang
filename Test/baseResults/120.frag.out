ERROR: 0:9: 'in for stage inputs' : not supported for this version or the enabled extensions 
ERROR: 0:10: 'out for stage outputs' : not supported for this version or the enabled extensions 
ERROR: 0:54: '+' :  wrong operand types: no operation '+' exists that takes a left-hand operand of type '2-component vector of float' and a right operand of type '3-component vector of float' (or there is no acceptable conversion)
ERROR: 0:55: '*' :  wrong operand types: no operation '*' exists that takes a left-hand operand of type 'uniform 4X2 matrix of float' and a right operand of type '3-component vector of float' (or there is no acceptable conversion)
ERROR: 0:56: '+' :  wrong operand types: no operation '+' exists that takes a left-hand operand of type 'uniform 4X2 matrix of float' and a right operand of type 'smooth in 4-component vector of float' (or there is no acceptable conversion)
ERROR: 0:57: '=' :  cannot convert from 'const float' to 'int'
ERROR: 0:58: 'assign' :  cannot convert from 'bool' to 'float'
ERROR: 0:59: '+' :  wrong operand types: no operation '+' exists that takes a left-hand operand of type 'int' and a right operand of type 'bool' (or there is no acceptable conversion)
ERROR: 0:60: '*' :  wrong operand types: no operation '*' exists that takes a left-hand operand of type 'bool' and a right operand of type 'int' (or there is no acceptable conversion)
ERROR: 0:60: 'assign' :  cannot convert from 'bool' to 'float'
ERROR: 0:61: 'assign' :  cannot convert from 'int' to 'bool'
ERROR: 0:62: '+' :  wrong operand types: no operation '+' exists that takes a left-hand operand of type 'bool' and a right operand of type 'float' (or there is no acceptable conversion)
ERROR: 0:63: 'bitwise-or assign' : not supported for this version or the enabled extensions 
ERROR: 0:63: 'assign' :  cannot convert from 'bool' to 'float'
ERROR: 0:79: ':' :  wrong operand types: no operation ':' exists that takes a left-hand operand of type '4-component vector of float' and a right operand of type '4X4 matrix of float' (or there is no acceptable conversion)
ERROR: 0:79: 'assign' :  cannot convert from '4X4 matrix of float' to 'fragColor 4-component vector of float'
ERROR: 0:82: 'xr' : illegal - vector component fields not from the same set 
ERROR: 0:83: 'z' : vector field selection out of range 
ERROR: 0:84: 'assign' :  l-value required 
ERROR: 0:90: 'int' : overloaded functions must have the same return type 
ERROR: 0:90: 'main' : function already has a body 
ERROR: 0:90: 'int' :  main function cannot return a value
ERROR: 0:91: 'main' : function cannot take any parameter(s) 
ERROR: 0:93: 'a' : variables with qualifier 'const' must be initialized 
ERROR: 0:96: 'out' : overloaded functions must have the same parameter qualifiers 
ERROR: 0:98: 'return' : function return is not matching type: 
ERROR: 26 compilation errors.  No code generated.

ERROR: node is still EOpNull!
0:21  Function Definition: main( (void)
0:21    Function Parameters: 
0:23    Sequence
0:23      Sequence
0:23        move second child to first child (2X3 matrix of float)
0:23          'm23' (2X3 matrix of float)
0:23          Construct mat2x3 (2X3 matrix of float)
0:23            'm' (uniform 4X2 matrix of float)
0:27      Sequence
0:27        move second child to first child (structure)
0:27          'sv' (structure)
0:27          Construct structure (structure)
0:27            Convert int to float (float)
0:27              'a' (int)
0:28      Sequence
0:28        move second child to first child (2-element array of float)
0:28          'ia' (2-element array of float)
0:28          Construct float (2-element array of float)
0:28            3.000000
0:28            direct index (float)
0:28              'i' (smooth in 4-component vector of float)
0:28              1 (const int)
0:29      Sequence
0:29        move second child to first child (float)
0:29          'f1' (float)
0:29          1.000000
0:30      Sequence
0:30        move second child to first child (float)
0:30          'f' (float)
0:30          Convert int to float (float)
0:30            'a' (int)
0:31      move second child to first child (float)
0:31        'f' (float)
0:31        Convert int to float (float)
0:31          'a' (int)
0:33      Sequence
0:33        move second child to first child (3-component vector of float)
0:33          'v3' (3-component vector of float)
0:33          Convert int to float (3-component vector of float)
0:33            'iv3' (3-component vector of int)
0:34      move second child to first child (float)
0:34        'f' (float)
0:34        add (float)
0:34          'f' (float)
0:34          Convert int to float (float)
0:34            'a' (int)
0:35      move second child to first child (float)
0:35        'f' (float)
0:35        subtract (float)
0:35          Convert int to float (float)
0:35            'a' (int)
0:35          'f' (float)
0:36      add second child into first child (float)
0:36        'f' (float)
0:36        Convert int to float (float)
0:36          'a' (int)
0:37      move second child to first child (float)
0:37        'f' (float)
0:37        subtract (float)
0:37          Convert int to float (float)
0:37            'a' (int)
0:37          'f' (float)
0:38      multiply second child into first child (3-component vector of float)
0:38        'v3' (3-component vector of float)
0:38        Convert int to float (3-component vector of float)
0:38          'iv3' (3-component vector of int)
0:39      move second child to first child (3-component vector of float)
0:39        'v3' (3-component vector of float)
0:39        divide (3-component vector of float)
0:39          Convert int to float (3-component vector of float)
0:39            'iv3' (3-component vector of int)
0:39          2.000000
0:40      move second child to first child (3-component vector of float)
0:40        'v3' (3-component vector of float)
0:40        vector-scale (3-component vector of float)
0:40          3.000000
0:40          Convert int to float (3-component vector of float)
0:40            'iv3' (3-component vector of int)
0:41      move second child to first child (3-component vector of float)
0:41        'v3' (3-component vector of float)
0:41        vector-scale (3-component vector of float)
0:41          2.000000
0:41          'v3' (3-component vector of float)
0:42      move second child to first child (3-component vector of float)
0:42        'v3' (3-component vector of float)
0:42        subtract (3-component vector of float)
0:42          'v3' (3-component vector of float)
0:42          2.000000
0:43      Test condition and select (void)
0:43        Condition
0:47        logical-or (bool)
0:46          logical-or (bool)
0:45            logical-or (bool)
0:44              logical-or (bool)
0:43                logical-or (bool)
0:43                  Compare Less Than (bool)
0:43                    'f' (float)
0:43                    Convert int to float (float)
0:43                      'a' (int)
0:44                  Compare Less Than or Equal (bool)
0:44                    Convert int to float (float)
0:44                      'a' (int)
0:44                    'f' (float)
0:45                Compare Greater Than (bool)
0:45                  'f' (float)
0:45                  Convert int to float (float)
0:45                    'a' (int)
0:46              Compare Greater Than or Equal (bool)
0:46                'f' (float)
0:46                Convert int to float (float)
0:46                  'a' (int)
0:47            Compare Equal (bool)
0:47              Convert int to float (float)
0:47                'a' (int)
0:47              'f' (float)
0:48          Compare Not Equal (bool)
0:48            'f' (float)
0:48            Convert int to float (float)
0:48              'a' (int)
0:43        true case is null
0:49      move second child to first child (float)
0:49        'f' (float)
0:49        Test condition and select (float)
0:49          Condition
0:49          'b' (bool)
0:49          true case
0:49          Convert int to float (float)
0:49            'a' (int)
0:49          false case
0:49          'f' (float)
0:50      move second child to first child (float)
0:50        'f' (float)
0:50        Test condition and select (float)
0:50          Condition
0:50          'b' (bool)
0:50          true case
0:50          'f' (float)
0:50          false case
0:50          Convert int to float (float)
0:50            'a' (int)
0:51      move second child to first child (float)
0:51        'f' (float)
0:51        Convert int to float (float)
0:51          Test condition and select (int)
0:51            Condition
0:51            'b' (bool)
0:51            true case
0:51            'a' (int)
0:51            false case
0:51            'a' (int)
0:52      Sequence
0:52        move second child to first child (structure)
0:52          'news' (structure)
0:52          'sv' (structure)
0:54      vector swizzle (2-component vector of float)
0:54        'i' (smooth in 4-component vector of float)
0:54        Sequence
0:54          0 (const int)
0:54          1 (const int)
0:55      'm' (uniform 4X2 matrix of float)
0:56      'm' (uniform 4X2 matrix of float)
0:58      'f' (float)
0:59      move second child to first child (float)
0:59        'f' (float)
0:59        Convert int to float (float)
0:59          'a' (int)
0:60      'f' (float)
0:61      'b' (bool)
0:62      move second child to first child (bool)
0:62        'b' (bool)
0:62        'b' (bool)
0:63      'f' (float)
0:65      move second child to first child (4-component vector of float)
0:65        'gl_FragColor' (fragColor 4-component vector of float)
0:65        Function Call: texture2D(s21;vf2; (4-component vector of float)
0:65          's2D' (uniform sampler2D)
0:65          'centTexCoord' (centroid smooth in 2-component vector of float)
0:?       Sequence
0:79        'gl_FragColor' (fragColor 4-component vector of float)
0:82      direct index (float)
0:82        'gl_FragColor' (fragColor 4-component vector of float)
0:82        0 (const int)
0:83      direct index (float)
0:83        'centTexCoord' (centroid smooth in 2-component vector of float)
0:83        0 (const int)
0:84      move second child to first child (bool)
0:84        Comma (bool)
0:84          'a' (int)
0:84          'b' (bool)
0:84        true (const bool)
0:90  Function Definition: main( (int)
0:90    Function Parameters: 
0:91  Function Definition: main(i1; (void)
0:91    Function Parameters: 
0:91      'a' (in int)
0:96  Function Definition: foo(f1; (int)
0:96    Function Parameters: 
0:96      'a' (out float)
0:98    Sequence
0:98      Branch: Return with expression
0:98        3.200000
0:99      Function Call: foo(f1; (int)
0:99        'a' (out float)
0:?   Linker Objects
0:?     'i' (smooth in 4-component vector of float)
0:?     'o' (out 4-component vector of float)
0:?     's2D' (uniform sampler2D)
0:?     'centTexCoord' (centroid smooth in 2-component vector of float)
0:?     'm' (uniform 4X2 matrix of float)

